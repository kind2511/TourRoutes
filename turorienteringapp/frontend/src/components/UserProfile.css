/*  background of the user profile component. */
.user-profile-background {
    display: flex;
    justify-content: center;
    align-items: center;
    height: 100vh;
}

/*  user profile form container. */
.user-profile-container {
    width: 500px;
    padding: 60px 60px 60px 40px; /* top right bottom left   */
    border-radius: 10px;
    box-shadow: 0px 0px 15px rgba(19, 182, 38, 0.937);
    background-color: #d5d2d2;
    margin-top: 50px; 
}

/*  User profile desing  */
h2 {
    margin-bottom: 28px;
    text-align: center;
    color: #333;
    font-weight: bold;
    font-size: 30px;
}

.edit-dropdown {
    position: relative;
    display: inline-block;
    margin-left: 15px;
}

.edit-dropdown-content {
    display: none;
    position: absolute;
    right: 0;
    background-color: #f9f9f9;
    min-width: 100px;
    box-shadow: 0px 0px 15px rgba(51, 240, 4, 0.1);
    z-index: 1;
    cursor: pointer;
}

/* style of the type on the edit button */
.edit-dropdown-content a {
    color: rgb(24, 23, 23);
    padding: 12px 16px;
    text-decoration: none;
    display: block;
}

/* hover of the edit button*/
.edit-dropdown-content a:hover {
    background-color: #0cedf1
}

/* drop down edit word when set the mouse curser*/
.edit-dropdown:hover .edit-dropdown-content {
    display: block;
}


.userprofile-logo{
    font-family: 'Pacifico', cursive;
    font-size: 38px;
    font-weight: bold;
    color: #4CAF50;
    letter-spacing: 1px;
    text-transform: capitalize;
    cursor: pointer;
     margin-top: -5px;
     margin-left: -10px;  
     text-align: center;

}