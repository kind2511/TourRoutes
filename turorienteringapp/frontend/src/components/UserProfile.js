import React, { useState, useEffect } from "react";
import { useNavigate } from "react-router-dom";
import "./UserProfile.css";

const UserProfile = () => {
  // State to store user's profile data
  const [userData, setUserData] = useState({});

  // State to manage the edit mode for each individual field
  const [isEditing, setIsEditing] = useState({
    firstName: false,
    lastName: false,
    email: false,
  });

  const navigate = useNavigate(); // Navigation function

  // Sends a get request to the backend and retrives the users personal data
  useEffect(() => {
    const fetchUserData = async () => {
      try {
        const response = await fetch(
          "http://localhost:8000/api/v1/users/myProfile",
          {
            method: "GET",
            headers: {
              "content-type": "application/json",
              Authorization: "Bearer " + localStorage.getItem("token"),
            },
            credentials: "same-origin",
          }
        );

        if (!response.ok) {
          throw new Error("Network response was not ok");
        }

        const data = await response.json();

        setUserData({
          firstName: data.data.user.firstName,
          lastName: data.data.user.lastName,
          email: data.data.user.email,
        });
      } catch (error) {
        console.error("Error fetching user data:", error);
      }
    };

    fetchUserData();
  }, []);

  const handleLogoClick = () => {
    navigate("/dashboard");
  };

  /**
   * Update a user's profile data and reset the edit mode for the field.
   */
  const handleUpdateProfile = async (field) => {
    setIsEditing((prevEditingState) => ({
      ...prevEditingState,
      [field]: false,
    }));

    // Sends the updated userdata to the backend via a patch request, which then persists the updated userdata.
    try {
      const response = await fetch(
        "http://localhost:8000/api/v1/users/updateMyProfile",
        {
          method: "PATCH",
          headers: {
            "content-type": "application/json",
            Authorization: "Bearer " + localStorage.getItem("token"),
          },
          credentials: "same-origin",
          body: JSON.stringify(userData),
        }
      );
    } catch (error) {
      console.error("Error updating user data:", error);
    }
  };

  //--------------------------------------------------------------------------------------
  const toggleEditing = (field) => {
    setIsEditing((prev) => ({ ...prev, [field]: !prev[field] }));
  };

  const handleDeleteClick = () => {
    navigate("/delete-user"); // Route to the DeleteUser component
  };
  //----------------------------------------------------------------------------------------

  // /*
  //  * Render an editable field with options to save or switch to edit mode.
  //  *
  //  * @param {string} field - The name of the field (e.g., "firstName")
  //  * @param {string} value - The current value of the field
  //  * @return {JSX.Element} - A React element that represents the editable field
  //  */
  const renderEditableField = (field, value) => (
    <div className="editable-field">
      <span>{value}</span>
      {isEditing[field] ? (
        <>
          <input
            value={value}
            onChange={(e) =>
              setUserData((prev) => ({ ...prev, [field]: e.target.value }))
            }
          />
          <button onClick={() => handleUpdateProfile(field)}>Save</button>
        </>
      ) : (
        <div className="edit-dropdown">
          <span>Edit</span>
          <div className="edit-dropdown-content">
            <a onClick={() => toggleEditing(field)}>Edit</a>
          </div>
        </div>
      )}
    </div>
  );

  {
    /*Render*/
  }
  return (
    <div className="user-profile-background">
      <div className="user-profile-container">
        <div className="userprofile-logo" onClick={handleLogoClick}>
          TurRuter
        </div>
        <p>
          <strong>First Name:</strong>
          {renderEditableField("firstName", userData.firstName)}
        </p>
        <p>
          <strong>Last Name:</strong>
          {renderEditableField("lastName", userData.lastName)}
        </p>
        <p>
          <strong>Email:</strong>
          {renderEditableField("email", userData.email)}
        </p>
        <div className="profile-action">
          {/* Changed from button to clickable text */}
          <span className="delete-user-text" onClick={handleDeleteClick}>
            Delete User
          </span>
        </div>
      </div>
    </div>
  );
};

export default UserProfile;
