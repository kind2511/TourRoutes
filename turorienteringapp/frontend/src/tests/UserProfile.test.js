import React from 'react';
import { render } from '@testing-library/react';
import UserProfile from '../components/UserProfile';
import { BrowserRouter } from 'react-router-dom';

test('UserProfile component renders without crashing', () => {
  render(
    <BrowserRouter>
      <UserProfile />
    </BrowserRouter>
  );
});
