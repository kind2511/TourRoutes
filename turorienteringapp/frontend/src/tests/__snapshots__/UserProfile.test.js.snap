// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserProfile component snapshot 1`] = `
<DocumentFragment>
  <div
    class="user-profile-background"
  >
    <button
      class="UserProfileBackButton"
    >
      Back
    </button>
    <div
      class="user-profile-container"
    >
      <div
        class="userprofile-logo"
      >
        TurRuter
      </div>
      <div>
        <strong>
          First Name:
        </strong>
        <div
          class="editable-field"
        >
          <div
            class="field-value-edit"
          >
            <span
              class="field-value"
            />
            <span
              class="edit-text"
            >
              Edit
            </span>
          </div>
        </div>
      </div>
      <div>
        <strong>
          Last Name:
        </strong>
        <div
          class="editable-field"
        >
          <div
            class="field-value-edit"
          >
            <span
              class="field-value"
            />
            <span
              class="edit-text"
            >
              Edit
            </span>
          </div>
        </div>
      </div>
      <div>
        <strong>
          Email:
        </strong>
        <div
          class="editable-field"
        >
          <div
            class="field-value-edit"
          >
            <span
              class="field-value"
            />
            <span
              class="edit-text"
            >
              Edit
            </span>
          </div>
        </div>
      </div>
      <div
        class="profile-action"
      >
        <span
          class="change-password-text"
        >
          Change Password
        </span>
        <span
          class="delete-user-text"
        >
          Delete User
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
